[#158825291] Make CodiMD deployable to Cloud Foundry

henrytk opened the pull request:
## What

Adapt this repository to work with Cloud Foundry.

The changes made here are not suitable to be merged upstream. This was a firebreak project, and therefore did the minimum implementation necessary.

## How to review

I failed to get this running on my local machine due to a weird issue with Webpack failing to build properly. However, you can test this directly on a dev environment:

```bash
cf create-service postgres tiny-unencrypted-9.5 hackmd-db
cf push --no-start
cf bind-service hackmd hackmd-db
cf start hackmd # when the database is ready
```

## Who

Anyone but me

henrytk committed:
Remove legacy config

These produced the following warning output on `cf push`:

```
warn: config.js contains deprecated lowercase setting for dbURL. Please
change your config.js file to replace dburl with dbURL
```

henrytk committed:
Add .cfignore

This is a copy of .gitignore, but with yarn.lock added.

It is unclear how or why Yarn is used alongside NPM.

If yarn.lock is pushed you receive the error:

```
Lockfile does not contain pattern "cfenv@^1.1.0"
```

Presumably because I would have had to recreate yarn.lock after adding
cfenv to the package.json. However, I did not do this because for Cloud
Foundry deployments the dependencies and compilation steps happen during
the staging process, so lock files are not required.

henrytk committed:
Add Cloud foundry manifest

Two issues:

1. High availability
2. Postgres client connectivity

1. High availability

Currently not possible to run multiple instances of this app.  The app
will start, but Javascript console shows connections to
wss://app-domain/socket.io/ are failing.  Presumably, the app is
stateful in some way. Example of the full error (secrets no longer
relevant):

```
Error: "xhr poll error"
onErrorhttps://cdnjs.cloudflare.com/ajax/libs/socket.io/2.0.4/socket.io.js:2:7328doPollhttps://cdnjs.cloudflare.com/ajax/libs/socket.io/2.0.4/socket.io.js:2:1979emithttps://cdnjs.cloudflare.com/ajax/libs/socket.io/2.0.4/socket.io.js:1:12370onErrorhttps://cdnjs.cloudflare.com/ajax/libs/socket.io/2.0.4/socket.io.js:2:3683onreadystatechangehttps://cdnjs.cloudflare.com/ajax/libs/socket.io/2.0.4/socket.io.js:2:3309 3.index.8e11967f017261b5c7b5.js:57463:4

Firefox can't establish a connection to the server at wss://paas-hackmd.henry.dev.cloudpipelineapps.digital/socket.io/?noteId=J0s7FWPWRliPtC7wZy6E6Q&EIO=3&transport=websocket&sid=2HXaCVYnYSszMDM4AAAF. websocket.js:112

Source map error: request failed with status 404
Resource URL: https://paas-hackmd.henry.dev.cloudpipelineapps.digital/build/3.index.8e11967f017261b5c7b5.js
Source Map URL: sequence-diagram-min.js.map[Learn More]
```

I thought it was the absence of a configured session secret, because
a random one gets generated if you don't specify one. However, setting
`CMD_SESSION_SECRET` in the environment produced the same results.

2. Postgres client connectivity

I set `PGSSLMODE` to `require` because otherwise the app logs an error
about failing to connect. The error message ended in `SSL OFF`. The
quick fix is to add this environment variable. A better long-term fix
would be to use a sensibly configured client.

henrytk committed:
Add cfenv

This allows easy parsing of VCAP_SERVICES data.

henrytk committed:
Make package.json work with Cloud Foundry

This does three things:

1. Pins newer Node and NPM versions
2. Runs `npm run build` instead of the Heroku build script.
3. Moves devDependencies into dependencies. These are required for
   the build script, so it makes no sense for them to be in the
   devDependencies section, because production apps NEED TO BE BUILT.
   These sections were being treated as build-time and run-time
   dependencies. I am not sure if this was a misuse of NPM or whether
   NPM inherently conflates the relationship of dev and prod with that
   of build-time and run-time dependencies. You decide.

henrytk committed:
Prevent webpack from hanging during staging

Staging was hanging indefinitely when it started optimizing assets
(part of the `npm run build` postinstall script).

I didn't investigate fully, but it seems to be related to Node's poor
grasp of system resources[1,2,3].

One suspect was identified in the build output:

```
> npm run build (node:46) MaxListenersExceededWarning: Possible
  EventEmitter memory leak detected. 11 SIGINT listeners added. Use
  emitter.setMaxListeners() to increase limit
```

[1] https://github.com/webpack/webpack/issues/2012#issuecomment-337263767
[2] https://github.com/webpack/webpack/issues/4558
[3] https://github.com/webpack-contrib/sass-loader/issues/100

alext commented on manifest.yml:
```diff
@@ -0,0 +1,11 @@
+    PGSSLMODE: require
+    UV_THREADPOOL_SIZE: 100
+    CMD_PORT: 8080
+    CF_DB: hackmd-db
```
I think specifying the DB name really belongs in the deployment pipeline as that's what creates/names the database.

henrytk replied:
I think you are right. I avoided it because it would have involved using some scripting to change the manifest within the Concourse task. I can change it if you want me to.

henrytk committed:
Remove deployment-time variables from manifest

This could vary between deployments, so should be injected at deploy
time in our pipelines.
